src/java.base/share/native/libjli/parse_manifest.c

```diff
@@ -140,6 +140,83 @@ readAt(int fd, jlong pos, unsigned int count, void *buf) {
             && read(fd, buf, count) == (jlong) count);
 }
 
+/*
+ * Reads size and offset fields from a zip64 extended information extra field.
+ *
+ * If one of the corresponding central directory fields is too small to hold the
+ * required data, the information will be in the zip64 extended information
+ * record.
+ *
+ * The disk number field is only used to validate the size of the zip64 extended
+ * information record.
+ *
+ * See APPNOTE.TXT 4.5.3.
+ */
+static jboolean
+read_zip64_ext(Byte *p, jlong *cenlen, jlong *censiz, jlong *cenoff,
+               unsigned short cendsk) {
+  short headerId = ZIPEXT_HDR(p);
+  if (headerId != ZIP64_EXTID) {
+    return JNI_FALSE;
+  }
+  short headerSize = ZIPEXT_SIZ(p);
+  short expectedSize =
+      (*cenlen == ZIP64_MAGICVAL ? 8 : 0) +
+      (*censiz == ZIP64_MAGICVAL ? 8 : 0) +
+      (*cenoff == ZIP64_MAGICVAL ? 8 : 0) +
+      (cendsk == ZIP64_MAGICCOUNT ? 4 : 0);
+  if (headerSize != expectedSize) {
+    return JNI_FALSE;
+  }
+  jlong offset = 4;
+  if (*cenlen == ZIP64_MAGICVAL) {
+    *cenlen = LL(p, offset);
+    offset += 8;
+  }
+  if (*censiz == ZIP64_MAGICVAL) {
+    *censiz = LL(p, offset);
+    offset += 8;
+  }
+  if (*cenoff == ZIP64_MAGICVAL) {
+    *cenoff = LL(p, offset);
+  }
+  return JNI_TRUE;
+}
+
+/**
+ * Validates the associated LOC headers for a CEN entry, with support for
+ * reading the LOC offset using the zip64 extended information extra field.
+ */
+static jboolean
+validate_lochdr(int fd, jlong censtart, jlong base_offset, Byte *cenhdr) {
+  Byte lochdr[LOCHDR];
+  jlong cenlen = CENLEN(cenhdr);
+  jlong censiz = CENSIZ(cenhdr);
+  jlong cenoff = CENOFF(cenhdr);
+  jlong cenext = CENEXT(cenhdr);
+  if (cenoff == ZIP64_MAGICVAL && cenext > 0) {
```

Comment:
Probably also need to check if `cenlen` or `censiz` is ZIP64_MAGICVAL?

Reply:
I think it doesn't matter, because the validation below only uses `cenoff`. If `cenoff` fits in 32 bits, we don't need to read the zip64 extra info.

Reply:
Thanks for the explanation. Could you please add a comment with the info above the `if` statement?

Reply:
I added a comment